lib/Dialect/QUIR/Transforms/BreakReset.cpp

```diff
@@ -131,13 +143,190 @@ void BreakResetPass::runOnOperation() {
+    auto newCallGateOp =
+        circuitBuilder.create<CallGateOp>(callGateOp.getLoc(), StringRef("x"),
+                                          TypeRange{}, circOp.getArguments());
+    finishCircuit(circOp);
+    resetGateCircuitsSymbolMap[llvm::StringRef(circuitName)] =
+        circOp.getOperation();
+  } else
+    circOp = dyn_cast<CircuitOp>(search->second);
+
+  mlir::OpBuilder builder(callGateOp);
+  auto callCircOp = builder.create<mlir::quir::CallCircuitOp>(
+      callGateOp->getLoc(), circOp.getSymName(), TypeRange{},
+      callGateOp.getOperands());
+
+  callCircOp->moveBefore(callGateOp);
+  callGateOp->erase();
+}
+
+void BreakResetPass::putMeasureInCircuit(ModuleOp moduleOp,
+                                         mlir::quir::MeasureOp measureOp) {
+  std::set<uint32_t> measureQubits =
+      QubitOpInterface::getOperatedQubits(measureOp);
+  ;
+  std::string circuitName = "reset_measure";
+  for (auto qubit : measureQubits)
+    circuitName += ("_" + std::to_string(qubit));
+
+  mlir::OpBuilder builder(measureOp);
+  std::vector<mlir::Type> const typeVec(measureOp.getQubits().size(),
+                                        builder.getI1Type());
+  auto search = resetGateCircuitsSymbolMap.find(circuitName);
+  CircuitOp circOp;
+  if (search == resetGateCircuitsSymbolMap.end()) {
+    // build a circuit
+    circOp = startCircuit<MeasureOp>(moduleOp, circuitName, measureOp);
+    OpBuilder circuitBuilder =
+        OpBuilder::atBlockBegin(&circOp.getBody().front());
+    auto resetMeasureOp = circuitBuilder.create<MeasureOp>(
+        measureOp.getLoc(), TypeRange(typeVec), circOp.getArguments());
+    resetMeasureOp->setAttr(getNoReportRuntimeAttrName(),
+                            builder.getUnitAttr());
+    finishCircuit(circOp);
+    resetGateCircuitsSymbolMap[llvm::StringRef(circuitName)] =
+        circOp.getOperation();
+  } else
+    circOp = dyn_cast<CircuitOp>(search->second);
+
+  auto callCircOp = builder.create<mlir::quir::CallCircuitOp>(
+      measureOp->getLoc(), circOp.getSymName(), TypeRange(typeVec),
+      measureOp.getOperands());
+
+  callCircOp->moveBefore(measureOp);
+  measureOp->replaceAllUsesWith(callCircOp);
+
+  measureOp->erase();
+}
+
+template <class measureOrCallGate>
+CircuitOp BreakResetPass::startCircuit(ModuleOp moduleOp,
+                                       const std::string &circuitName,
+                                       measureOrCallGate quantumGate) {
+  mlir::func::FuncOp mainFunc =
+      dyn_cast<mlir::func::FuncOp>(quir::getMainFunction(moduleOp));
+  assert(mainFunc && "could not find the main func");
+  mlir::OpBuilder builder(mainFunc);
+  mlir::Location location = mainFunc.getLoc();
+
+  // mangle the circuit name if there exist a circuit with the same name in
+  // input of this pass
+  auto mangledCircuitName = circuitName;
+  int cnt = 0;
+  while (inputCircuitsSymbolMap.contains(mangledCircuitName))
+    mangledCircuitName += std::to_string(cnt++);
```

Comment:
Do we have any intuition of how many times this is going to occur? 

I'm afraid this is going to end up being very expensive if we end up searching the `inputCircuitsSymbolMap` frequently). As a future improvement, it would be great if we could find a name mangling mechanism that is O(1) by appending a sufficiently long random string to the end. We may also be able to adopt some techniques from the C++ name mangling mechanisms.